rust/chains/hyperlane-cosmos/src/interchain_security_module.rs

```diff
@@ -0,0 +1,108 @@
+pub struct CosmosInterchainSecurityModule {
+    /// The domain of the ISM contract.
+    domain: HyperlaneDomain,
+    /// The address of the ISM contract.
+    address: H256,
+    /// The provider for the ISM contract.
+    provider: Box<WasmGrpcProvider>,
+}
+
+/// The Cosmos Interchain Security Module Implementation.
+impl CosmosInterchainSecurityModule {
+    /// Creates a new Cosmos Interchain Security Module.
+    pub fn new(conf: &ConnectionConf, locator: ContractLocator, signer: Signer) -> Self {
+        let provider: WasmGrpcProvider =
+            WasmGrpcProvider::new(conf.clone(), locator.clone(), signer.clone());
+
+        Self {
+            domain: locator.domain.clone(),
+            address: locator.address,
+            provider: Box::new(provider),
+        }
+    }
+}
+
+impl HyperlaneContract for CosmosInterchainSecurityModule {
+    fn address(&self) -> H256 {
+        self.address
+    }
+}
+
+impl HyperlaneChain for CosmosInterchainSecurityModule {
+    fn domain(&self) -> &HyperlaneDomain {
+        &self.domain
+    }
+
+    fn provider(&self) -> Box<dyn HyperlaneProvider> {
+        Box::new(CosmosProvider::new(self.domain.clone()))
+    }
+}
+
+fn ism_type_to_module_type(ism_type: hpl_interface::ism::ISMType) -> ModuleType {
+    match ism_type {
+        hpl_interface::ism::ISMType::Unused => ModuleType::Unused,
+        hpl_interface::ism::ISMType::Routing => ModuleType::Routing,
+        hpl_interface::ism::ISMType::Aggregation => ModuleType::Aggregation,
+        hpl_interface::ism::ISMType::LegacyMultisig => ModuleType::MessageIdMultisig,
+        hpl_interface::ism::ISMType::MerkleRootMultisig => ModuleType::MerkleRootMultisig,
+        hpl_interface::ism::ISMType::MessageIdMultisig => ModuleType::MessageIdMultisig,
+        hpl_interface::ism::ISMType::Null => ModuleType::Null,
+        hpl_interface::ism::ISMType::CcipRead => ModuleType::CcipRead,
+    }
+}
+
+#[async_trait]
+impl InterchainSecurityModule for CosmosInterchainSecurityModule {
+    /// Returns the module type of the ISM compliant with the corresponding
+    /// metadata offchain fetching and onchain formatting standard.
+    async fn module_type(&self) -> ChainResult<ModuleType> {
+        let query = QueryIsmModuleTypeRequest {
+            module_type: EmptyStruct {},
+        };
+
+        let data = self
+            .provider
+            .wasm_query(QueryIsmGeneralRequest { ism: query }, None)
+            .await?;
+
+        // Handle both the ISMType response and the ModuleTypeResponse response.
+        let ismtype_response = serde_json::from_slice::<QueryIsmModuleTypeResponse>(&data);
+        let moduletye_response =
+            serde_json::from_slice::<hpl_interface::ism::ModuleTypeResponse>(&data);
+
+        Ok(match (ismtype_response, moduletye_response) {
+            (Ok(v), _) => ism_type_to_module_type(v.typ),
+            (_, Ok(v)) => ism_type_to_module_type(v.typ),
+            _ => ModuleType::Null,
+        })
```

Comment:
I would first assign this to a variable, say `module_type`, and then return `Ok(module_type)`. Also, although the `match` here is more concise, to me this logic is easier to understand when written as
```suggestion
                let module_type = if let Ok(ism_type) = ismtype_response {
            ism_type_to_module_type(ism_type.typ)
        } else if let Ok(module_type) = moduletye_response {
            ism_type_to_module_type(module_type.typ)
        } else {
            ModuleType::Null
        };
        Ok(module_type)
```